src/lifetimes.md

```diff
@@ -213,3 +213,205 @@ totally ok*, because it keeps us from spending all day explaining our program
+we know that  a `&'s1 str'` outlives an `&'s2 str`, so we can substitute `&'s1
+s1` with `&'s2 s2`. After this both arguments are of lifetime `&'s2` and the
+call-site is consistent with the signature of `print_shortest`.
+
+> More formally, the basis for the above rule is in *type variance*. Under this
+> model, you would consider a longer lifetime a sub-type of a shorter lifetime,
+> and for function arguments to be *co-variant*. However, an understanding of
+> variance isn't strictly required to understand the Rust borrow checker. We've
+> tried here to instead to explain using intuitive terminlolgy.
+
+## Inter-procedural Borrow Checking of Function Return Values
+
+Now consider a slight variation of this example:
+
+```rust,ignore
+fn main() {
+    let s1 = String::from("short");
+    let res;
+    let s2 = String::from("a long long long string");
+    res = shortest(&s1, &s2);
+    println!("{}", res);
+}
+
+fn shortest<'k>(x: &'k str, y: &'k str) -> &'k str {
+    if x.len() < y.len() {
+        return x;
+    } else {
+        return y;
+    }
+}
+```
+
+`print_shortest` has been renamed to `shortest`, which instead of printing,
+now returns the shorter of the two strings. It does this using only references
+for efficiency, avoiding the need to re-allocate a new string to pass back to
+`main`. The responsibility of printing the result has been shifted to `main`.
+
+Let's again de-sugar `main` by adding explicit scopes and lifetimes:
+
+```rust,ignore
+fn main() {
+    's1 {
+        let s1 = String::from("short");
+        'res {
+            let res: &'res str;
+            's2 {
+                let s2 = String::from("a long long long string");
+                res: &'res: str = shortest(&'s1 s1, &'s2 s2);
+                println!("{}", res);
+            }
+        }
+    }
+}
+```
+
+Again, at the call-site of `shortest` the compiler needs to check the
+consistency of the arguments in the caller with the signature of the callee.
+The signature of `shortest` first says that the two reference arguments have
+the same lifetime, which can be prove ok in the same way as before, thus giving
+us:
+
+```rust,ignore
+res: &'res = shortest(&'s2 s1, &'s2 s2);
+```
+
+But we now have the additional reference to check. We must now prove that the
+returned reference can have the same lifetime as the arguments of lifetime
+`&'s2`. This brings us to a second rule:
+
+> The return value of a function `&'r T` can be converted to an argument `&'s T`
+> if the lifetime of `&'r T` is equal or shorter than `&'s T`.
+
+To make our program compile, we would have to subsitute `res: &'res` for `res:
+&'s2`, but we can't since `&'res` in fact out-lives `&'s2`. This program is
+inconsistent and the compiler rightfully rejects the program because we
+try make a reference (`res`) which outlives one of the values it may refer to
+(`s2`).
+
+> Formally, function return values are said to be *contravariant*, the opposite
+> of *covariant*.
```

Comment:
Here's my concrete suggestion for this PR:

1. Rewrite Rule I (L289-290) to say something along the lines of "A reference can always be shrunk to one of a shorter lifetime.  In other words, `&'a T` can be implicitly converted to `&'b T` as long as `'a` outlives `'b`."

2. Replace "for function arguments to be co-variant" (L299) with "a reference `&'a T` is always covariant in its lifetime parameter `'a`."

3. Replace Rule II and adjacent paragraphs (L348-372) with something like:

    > Again, at the call-site of `shortest` the compiler needs to check the consistency of the arguments in the caller with the signature of the callee.  The signature of `shortest` says that all three references must have the same lifetime `'k`, so we have to find a lifetime `'k` such that:
    >
    >   - `&s1: &'s1 str` can be converted to the first argument `&'k str`,
    >   - `&s2: &'s2 str` can be converted to the second argument `&'k str`, and
    >   - the return value `&'k str` can be converted to `res: &'res str`
    >
    > ```rust,ignore
    > res: &'res = shortest::<'k>(&'s1 s1, &'s2 s2);
    >           //                ^^^^^^^
    >           //               's1 -> 'k ^^^^^^^
    >           //                        's2 -> 'k
    >           // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    >           //           'k -> 'res
    > ```
    >
    > This leads to three requirements:
    >
    >   - `'s1` must outlive `'k`, 
    >   - `'s2` must outlive `'k`, and
    >   - `'k` must outlive `'res`.
    >
    > However,  `'res` *strictly outlives* `'s2` by virtue of being a larger scope that encloses `s2`, so it's impossible to satisfy these conditions.  The borrow checker rightfully rejects our program because we are making a reference (`res`) which outlives one of the values it may refer to (`s2`)."